app/ui/src/app/integration/edit-page/step-configure/filter-steps/basic-filter.component.ts

```diff
@@ -66,53 +61,106 @@ export class BasicFilterComponent implements OnChanges {
+    if (configuredProperties && configuredProperties.rules) {
+      // TODO hackity hack
+      if (typeof configuredProperties.rules === 'string') {
+        rules = JSON.parse(<any>configuredProperties.rules);
+      } else {
+        rules = configuredProperties.rules;
+      }
 
-    // this can be valid even if we can't fetch the form data
-    function initializeForm(ops?, paths?) {
-      self.basicFilterModel = createBasicFilterModel(
-        self.configuredProperties || <any>{},
-        ops,
-        paths
-      );
-      self.formGroup = self.dynamicFormService.createFormGroup(
-        self.basicFilterModel
-      );
-      self.predicateControl = self.formGroup
-        .get('filterSettingsGroup')
-        .get('predicate') as FormControl;
-      self.predicateModel = findById(
-        'predicate',
-        self.basicFilterModel
-      ) as DynamicInputModel;
-      self.rulesArrayControl = self.formGroup
-        .get('rulesGroup')
-        .get('rulesFormArray') as FormArray;
-      self.rulesArrayModel = findById(
-        'rulesFormArray',
-        self.basicFilterModel
-      ) as DynamicFormArrayModel;
-      self.loading = false;
-      self.valid = self.formGroup.valid;
-      self.validChange.emit(self.formGroup.valid);
+      for (const incomingRule of rules) {
+        incomingGroups.push(this.fb.group(incomingRule));
+      }
     }
 
+    this.ruleOpsModel$ = of(ops);
+    this.pathsModel = paths;
+
+    this.filterSettingsGroupObj = {
+      predicate: {
+        label: 'Continue only if incoming data match',
+        options: this.predicateOpsModel$,
+        value: (configuredProperties && configuredProperties.predicate) ? configuredProperties.predicate : 'AND'
+      }
+    };
+
+    let preloadedRulesArray;
+
+    if (incomingGroups.length > 0) {
+      preloadedRulesArray = this.fb.array(incomingGroups);
+    } else {
+      preloadedRulesArray = this.fb.array([this.createNewRuleGroup()]);
+    }
+
+    this.rulesArray = preloadedRulesArray;
+
+    const formGroupObj = {
+      filterSettingsGroup: this.fb.group(this.filterSettingsGroupObj),
+      rulesArray: preloadedRulesArray
+    };
+
+    this.form = this.fb.group(formGroupObj);
+
+    this.formValudChangeSubscription = this.form.valueChanges.subscribe(_ => {
+      this.valid = this.form.valid;
+      this.validChange.emit(this.valid);
+    });
+
+    this.loading = false;
+  }
+
+  ngOnDestroy() {
+    this.formValudChangeSubscription.unsubscribe();
+  }
+
+  ngOnChanges(changes: any) {
+    if (!('position' in changes)) {
+      return;
+    }
+
+    this.loading = true;
     // Fetch our form data
     this.integrationSupportService
       .getFilterOptions(this.dataShape)
       .toPromise()
       .then((body: any) => {
         const ops = body.ops;
         const paths = body.paths;
-        initializeForm(ops, paths);
+        this.ruleOpsModel$ = of(body.ops);
+        this.pathsModel = body.ops;
```

Comment:
This should be 'paths' shouldn't it?  This would eventually hold an array of strings that would be used as a typeahead for the 'path' part of a given rule in the ruleset array.